README.md
Based on [oischinger's eufyp2pstream addon](https://github.com/oischinger/eufyp2pstream).
# Goal and Progress:

## Goal
More stable experience in Homekit than Homebridge Eufy Security plugin.  Video skips frames, audio becomes unsynced.

More flexbility in RTSP url generation and manipulation for Eufy Doorbell using P2P stream.

Gstreamer pipeline.

If koush exposes two-way audio in Homekit in RTSP or FFmpeg cams I'll probably put talkback in.

## Progress
As of version 0.2 gst-launch-1.0 pipelines can be built and used in the go2rtc config.  Camera name must be left as default 'camera1' for now.  TCP port 63336 is used for video, 63337 for audio, 63338 is just for talkback (I think).  Config file is stored in /config/eufyp2p/go2rtc.yaml.  Default config works for my Eufy 2k Battery Doorbell, encoding has to be set to Low (Medium looks like it may work).

Options for custom snapshot interval don't currently work.  FFmpeg subprocess creates snapshot every 5 minutes in /config/www/eufyp2p/camera1.jpg which can be pulled from Home Assistant with http://<hass.ip>:<hass.port>/local/eufyp2p/camera1.jpg snapshot url.  On-demand snapshot generation just didn't really provide a stable experience.

## Other Codec/Homekit Info
Low/medium encoding uses h264 for video and high uses h265.  Low quality is 640x480, high is 1600x1200.  Have mine set to low quality, low encoding for now to try to start streams faster.

My stream is fed to Scrypted RTSP cam, ondemand without Rebroadcast prebuffer.  Startup in Homekit seems to be affected more by where stream gets started in relation to keyframes (keyframe interval is 4 seconds).  If it gets started just before a keyframe it'll start in ~4-6 seconds.  If it starts just after a keyframe it'll start in 8-11 seconds.  Seems to be a by-product of no prebuffer.  Going to try a new doorbell transformer to see if it can't keep up with the discharge rate to keep a prebuffer 24/7.

Below is info from oischinger's repo:

# EufyP2PStream

A small project that provides a Video/Audio Stream from Eufy cameras that don't directly support RTSP.

It uses [go2RTC](https://github.com/AlexxIT/go2rtc) to provide the live stream.

# Howto install

## Prerequisites
Install and configure the [Eufy Security Addon](https://github.com/fuatakgun/eufy_security_addon) first. See the instructions [here](https://github.com/fuatakgun/eufy_security).

## Install this addon
Copy this directory to `/addons/eufyp2pstream` on your Home Assistant host and [install the addon](https://my.home-assistant.io/redirect/supervisor_addon/?addon=local_eufyp2pstream).

Open the [Addon's WebUI](https://my.home-assistant.io/redirect/supervisor_ingress/?addon=local_eufyp2pstream) and enjoy a camera stream with audio via WEBRTC.

## Setting up a camera in Home Assistance

Go to devices and [setup a new generic camera](https://my.home-assistant.io/redirect/config_flow_start/?domain=generic) with the following parameters (replace `IP_ADDR` with your Home Assistant IP address):

- Static Image URL: `http://IP_ADDR:8787`
- Stream source: `rtsp://IP_ADDR:8541/camera1`

⚠️ I don't recommend using the Home Assistance camera integration at the moment since it seems to keep the Livestream ALWAYS active.

# References
This project is inspired by:

- [eufy_security Home Assistant Custom Component](https://github.com/fuatakgun/eufy_security)
- [go2RTC](https://github.com/AlexxIT/go2rtc)
- [Video Decoding in ioBroker.eusec](https://github.com/bropat/ioBroker.eusec/blob/0a15e1d125f4fd00144af66d57d8d738140ea619/src/lib/eufy-security/video.ts#L14-L65
)

# TODO

## Static Image:

A static image of the current stream can be obtained by heading to `http://IP_ADDR:8787`

## Talkback

To play an audio file start the P2P stream via the WebRTC URL and spin up an ffmpeg process like this:

`ffmpeg  -re -stream_loop -1 -i test.mp3 -vn -sn -dn -c:a aac  -b:a 20k -ar 16k -ac 2 -f adts tcp://127.0.0.1:63338`


CHANGELOG.md
## 0.5
- bump go2rtc to 1.2.0
- change snapshot interval from 10 to 30 min

## 0.4
- go2rtc to 1.1.2

## 0.3
- Change interval snapshot to rtsp instead of grabbing from websocket
- Use gstreamer to capture snapshot
- Set default snapshot to 10 min
- Bump go2rtc to 1.1.0
- Some dockfile optimizations and general cleanup

## 0.2
- Snapshot on an interval works, set to 5 minutes
- Custom interval steps created with options, not working currently

## 0.1
- First working version working as addon with gstreamer pipeline, no documentation


app/hello.py

from websocket import EufySecurityWebSocket
import aiohttp
import asyncio
import json
import socket
import threading
import time
import sys

import os
from queue import Queue

HOST_NAME = '0.0.0.0'

RTSP_PORT_NUMBER = 8541

RECV_CHUNK_SIZE = 4096

video_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
audio_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
backchannel_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

# Default camera name in go2rtc
CAMERA = 'camera1'
# Take snapshot every interval seconds
SSINTERVAL = 1800


EVENT_CONFIGURATION: dict = {
    "livestream video data": {
        "name": "video_data",
        "value": "buffer",
        "type": "event",
    },
    "livestream audio data": {
        "name": "audio_data",
        "value": "buffer",
        "type": "event",
    },
}

START_P2P_LIVESTREAM_MESSAGE = {
    "messageId": "start_livesteam",
    "command": "device.start_livestream",
    "serialNumber": None,
}

STOP_P2P_LIVESTREAM_MESSAGE = {
    "messageId": "stop_livesteam",
    "command": "device.stop_livestream",
    "serialNumber": None,
}

STOP_TALKBACK = {
    "messageId": "stop_talkback",
    "command": "device.stop_talkback",
    "serialNumber": None,
}

SET_API_SCHEMA = {
    "messageId": "set_api_schema",
    "command": "set_api_schema",
    "schemaVersion": 13,
}

STOP_P2P_LIVESTREAM_MESSAGE = {
    "messageId": "stop_livesteam",
    "command": "device.stop_livestream",
    "serialNumber": None,
}

P2P_LIVESTREAMING_STATUS = "p2pLiveStreamingStatus"

START_LISTENING_MESSAGE = {"messageId": "start_listening", "command": "start_listening"}

DRIVER_CONNECT_MESSAGE = {"messageId": "driver_connect", "command": "driver.connect"}

class ClientAcceptThread(threading.Thread):
    def __init__(self,socket,run_event,name,ws,serialno):
        threading.Thread.__init__(self)
        self.socket = socket
        self.queues = []
        self.run_event = run_event
        self.name = name
        self.ws = ws
        self.serialno = serialno
        self.my_threads = []

    def update_threads(self):
        my_threads_before = len(self.my_threads)
        for thread in self.my_threads:
            if not thread.is_alive():
                self.queues.remove(thread.queue)
        self.my_threads = [t for t in self.my_threads if t.is_alive()]
        if self.ws and my_threads_before > 0 and len(self.my_threads) == 0:
            if self.name == "BackChannel":
                print("All clients died. Stopping Stream for ", self.name)
            else:
                print("All clients died. Stopping Stream for ", self.name)

                msg = STOP_P2P_LIVESTREAM_MESSAGE.copy()
                msg["serialNumber"] = self.serialno
                asyncio.run(self.ws.send_message(json.dumps(msg)))

    def run(self):
        print("Accepting connection for ", self.name)
        msg = STOP_TALKBACK.copy()
        msg["serialNumber"] = self.serialno
        asyncio.run(self.ws.send_message(json.dumps(msg)))
        while self.run_event.is_set():
            self.update_threads()
            sys.stdout.flush()
            try:
                client_sock, client_addr = self.socket.accept()
                print ("New connection added: ", client_addr, " for ", self.name)
                sys.stdout.flush()

                if self.name == "BackChannel":
                    client_sock.setblocking(True)
                    print("Starting BackChannel")
                    thread = ClientRecvThread(client_sock, run_event, self.name, self.ws, self.serialno)
                    thread.start()
                else:
                    client_sock.setblocking(False)
                    thread = ClientSendThread(client_sock, run_event, self.name, self.ws, self.serialno)
                    self.queues.append(thread.queue)
                    if self.ws:
                        msg = START_P2P_LIVESTREAM_MESSAGE.copy()
                        msg["serialNumber"] = self.serialno
                        asyncio.run(self.ws.send_message(json.dumps(msg)))
                    self.my_threads.append(thread)
                    thread.start()
            except socket.timeout:
                pass

class ClientSendThread(threading.Thread):
    def __init__(self,client_sock,run_event,name,ws,serialno):
        threading.Thread.__init__(self)
        self.client_sock = client_sock
        self.queue = Queue(100)
        self.run_event = run_event
        self.name = name
        self.ws = ws
        self.serialno = serialno

    def run(self):
        try:
            while self.run_event.is_set():
                if self.queue.empty():
                    # Send something to know if socket is dead
                    self.client_sock.sendall(bytearray(0))
                    time.sleep(0.1)
                self.client_sock.sendall(
                    bytearray(self.queue.get(True)["data"])
                )
        except socket.error as e:
            print("Connection lost", self.name, e)
            pass
        except socket.timeout:
            print("Timeout on socket for ", self.name)
            pass
        self.client_sock.close()

class ClientRecvThread(threading.Thread):
    def __init__(self,client_sock,run_event,name,ws,serialno):
        threading.Thread.__init__(self)
        self.client_sock = client_sock
        self.run_event = run_event
        self.name = name
        self.ws = ws
        self.serialno = serialno

    def run(self):
        try:
            while self.run_event.is_set():
                try:
                    data = self.client_sock.recv(RECV_CHUNK_SIZE)
                    print("got ", len(data))
                except BlockingIOError:
                    # Resource temporarily unavailable (errno EWOULDBLOCK)
                    pass
        except socket.error as e:
            print("Connection lost", self.name, e)
            pass
        except socket.timeout:
            print("Timeout on socket for ", self.name)
            pass
        self.client_sock.close()
        msg = STOP_TALKBACK.copy()
        msg["serialNumber"] = self.serialno
        asyncio.run(self.ws.send_message(json.dumps(msg)))

class Connector:
    def __init__(
        self,
        run_event,
    ):
        video_sock.bind(("127.0.0.1", 63336))
        video_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        video_sock.settimeout(1) # timeout for listening
        video_sock.listen()
        audio_sock.bind(("127.0.0.1", 63337))
        audio_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        audio_sock.settimeout(1) # timeout for listening
        audio_sock.listen()
        backchannel_sock.bind(("127.0.0.1", 63338))
        backchannel_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        backchannel_sock.settimeout(1) # timeout for listening
        backchannel_sock.listen()
        self.ws = None
        self.run_event = run_event
        self.serialno = ""

    def setWs(self, ws : EufySecurityWebSocket):
        self.ws = ws

    async def on_open(self):
        print(f" on_open - executed")

    async def on_close(self):
        print(f" on_close - executed")

    async def on_error(self, message):
        print(f" on_error - executed - {message}")

    async def on_message(self, message):
        #print(f"- on_message {message}")
        payload = message.json()
        message_type: str = payload["type"]
        if message_type == "result":
            message_id = payload["messageId"]
            print(f"on_message - {payload}")
            if message_id == START_LISTENING_MESSAGE["messageId"]:
                message_result = payload[message_type]
                states = message_result["state"]
                for state in states["devices"]:
                    self.serialno = state["serialNumber"]
                self.video_thread = ClientAcceptThread(video_sock, run_event, "Video", self.ws, self.serialno)
                self.audio_thread = ClientAcceptThread(audio_sock, run_event, "Audio", self.ws, self.serialno)
                self.backchannel_thread = ClientAcceptThread(backchannel_sock, run_event, "BackChannel", self.ws, self.serialno)
                self.audio_thread.start()
                self.video_thread.start()
                self.backchannel_thread.start()

        if message_type == "event":
            message = payload[message_type]
            event_type = message["event"]
            if message["event"] == "livestream audio data":
                event_value = message[EVENT_CONFIGURATION[event_type]["value"]]
                event_data_type = EVENT_CONFIGURATION[event_type]["type"]
                if event_data_type == "event":
                    for queue in self.audio_thread.queues:
                        if queue.full():
                            # print("Video queue full.")
                            queue.get(False)
                        queue.put(event_value)
            if message["event"] == "livestream video data":
                event_value = message[EVENT_CONFIGURATION[event_type]["value"]]
                event_data_type = EVENT_CONFIGURATION[event_type]["type"]
                if event_data_type == "event":
                    for queue in self.video_thread.queues:
                        if queue.full():
                            # print("Video queue full.")
                            queue.get(False)
                        queue.put(event_value)
            if message["event"] == "livestream error":
                print("Livestream Error!")
                if self.ws and len(self.video_thread.queues) > 0:
                    msg = START_P2P_LIVESTREAM_MESSAGE.copy()
                    msg["serialNumber"] = self.serialno
                    asyncio.run(self.ws.send_message(json.dumps(msg)))


async def SnapshotInterval(CAM_NAME, INTERVAL):
    await asyncio.sleep(120)
    while True:
        try:
            os.system('ffmpeg -analyzeduration 1200000 -f h264 -i tcp://127.0.0.1:63336?timeout=100000000 -strict -2 -hls_init_time 0 -hls_time 2 -hls_segment_type mpegts -fflags genpts+nobuffer+flush_packets -frames:v 1 /config/www/eufyp2p/' + CAM_NAME + '.jpg')
        except Exception as e:
            print("Exception: ", e)

        await asyncio.sleep(INTERVAL)

async def main(run_event):
    c = Connector(run_event)

    ws: EufySecurityWebSocket = EufySecurityWebSocket(
        "127.0.0.1",
        3000,
        aiohttp.ClientSession(),
        c.on_open,
        c.on_message,
        c.on_close,
        c.on_error,
    )
    c.setWs(ws)
    await ws.connect()
    
    await ws.send_message(json.dumps(START_LISTENING_MESSAGE))
    await ws.send_message(json.dumps(SET_API_SCHEMA))
    await ws.send_message(json.dumps(DRIVER_CONNECT_MESSAGE))
    
    SnapshotInterval(CAMERA, SSINTERVAL)
    await asyncio.sleep(1000000000000000000000005)

run_event = threading.Event()
run_event.set()
try:
    asyncio.run(main(run_event))
except (KeyboardInterrupt, SystemExit):
    #httpd.server_close()
    run_event.clear()


docker/install-main-deps.sh
#!/bin/bash
set -e

RUN \
    apt-get update && \
    apt-get install -y --no-install-recommends \
        ca-certificates \
        libatk1.0-0 \
        libatk-bridge2.0-0 \
        libatk-adaptor \
        libatspi2.0-0 \
        libunwind8 \
        libdw1 \
        libgmp10 \
        libgsl27 \
        libglib2.0-0 \
        libcap2 \
        libcups2 \
        liborc-0.4-0 \
        iso-codes \
        libgl1 \
        libgles1 \
        libgles2 \
        libgudev-1.0-0 \
        libgbm1 \
        libgraphene-1.0-dev \
        libpng16-16 \
        libjpeg8 \
        libogg0 \
        libopus0 \
        libpango-1.0-0 \
        libvisual-0.4-0 \
        libtheora0 \
        libvorbis0a \
        libxkbcommon0 \
        libxcomposite1 \
        libxdamage1 \
        libwayland-client0 \
        libwayland-cursor0 \
        libwayland-egl1 \
        libwayland-server0 \
        libharfbuzz-icu0 \
        libegl1 \
        libepoxy0 \
        libgcrypt20 \
        libwebp7 \
        libwebpdemux2 \
        libwebpmux3 \
        libopenjp2-7 \
        libwoff1 \
        libxslt1.1 \
        bubblewrap \
        libseccomp2 \
        xdg-dbus-proxy \
        libsoup2.4-1 \
        libvulkan1 \
        libass9 \
        libchromaprint1 \
        libcurl3-gnutls \
        libaom3 \
        libbz2-1.0 \
        liblcms2-2 \
        libbs2b0 \
        libdca0 \
        libfaac0 \
        libfaad2 \
        libflite1 \
        libssl3 \
        ladspa-sdk \
        libfdk-aac2 \
        libgsm1 \
        libkate1 \
        libgme0 \
        libde265-0 \
        liblilv-0-0 \
        libmodplug1 \
        mjpegtools \
        libmjpegutils-2.1-0 \
        libmpcdec6 \
        libdvdnav4 \
        libdvdread8 \
        librsvg2-2 \
        librtmp1 \
        libsbc1 \
        libsndfile1 \
        libsoundtouch1 \
        libspandsp2 \
        libsrt1.4-openssl \
        libsrtp2-1 \
        libvo-aacenc0 \
        libvo-amrwbenc0 \
        libwebrtc-audio-processing1 \
        libofa0 \
        libzvbi0 \
        libopenexr25 \
        libwildmidi2 \
        libx265-199 \
        libzbar0 \
        wayland-protocols \
        libaa1 \
        libmp3lame0 \
        libcaca0 \
        libdv4 \
        libmpg123-0 \
        libvpx7 \
        libshout3 \
        libspeex1 \
        libtag1v5 \
        libtwolame0 \
        libwavpack1 \
        liba52-0.7.4 \
        libx264-163 \
        libopencore-amrnb0 \
        libopencore-amrwb0 \
        libmpeg2-4 \
        libavcodec58 \
        libavfilter7 \
        libavformat58 \
        libavutil56 \
        libva2 \
        libva-wayland2 \
        xvfb \
        libxrandr-dev \
        glibc-tools

root/app/eufyp2p/__init__.py
#!/usr/bin/env python3
from .websocket import EufySecurityWebSocket
import aiohttp
import asyncio
import json
import socket
import threading
import time
import sys
import subprocess
import os
from queue import Queue

HOST_NAME = '0.0.0.0'

RTSP_PORT_NUMBER = 8541

RECV_CHUNK_SIZE = 4096

video_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
audio_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
backchannel_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

# Default camera name in go2rtc
audioVideoStream = 'camera1'
# Take snapshot every interval in seconds
## set default to every 5 minutes
SSINTERVAL = 1800

ffmpegbin = "/usr/bin/ffmpeg"

EVENT_CONFIGURATION: dict = {
    "livestream video data": {
        "name": "video_data",
        "value": "buffer",
        "type": "event",
    },
    "livestream audio data": {
        "name": "audio_data",
        "value": "buffer",
        "type": "event",
    },
}

START_P2P_LIVESTREAM_MESSAGE = {
    "messageId": "start_livesteam",
    "command": "device.start_livestream",
    "serialNumber": None,
}

STOP_P2P_LIVESTREAM_MESSAGE = {
    "messageId": "stop_livesteam",
    "command": "device.stop_livestream",
    "serialNumber": None,
}

STOP_TALKBACK = {
    "messageId": "stop_talkback",
    "command": "device.stop_talkback",
    "serialNumber": None,
}

SET_API_SCHEMA = {
    "messageId": "set_api_schema",
    "command": "set_api_schema",
    "schemaVersion": 13,
}

STOP_P2P_LIVESTREAM_MESSAGE = {
    "messageId": "stop_livesteam",
    "command": "device.stop_livestream",
    "serialNumber": None,
}

P2P_LIVESTREAMING_STATUS = "p2pLiveStreamingStatus"

START_LISTENING_MESSAGE = {"messageId": "start_listening", "command": "start_listening"}

DRIVER_CONNECT_MESSAGE = {"messageId": "driver_connect", "command": "driver.connect"}

run_event = threading.Event()
run_event.set()

class ClientAcceptThread(threading.Thread):
    def __init__(self,socket,run_event,name,ws,serialno):
        threading.Thread.__init__(self)
        self.socket = socket
        self.queues = []
        self.run_event = run_event
        self.name = name
        self.ws = ws
        self.serialno = serialno
        self.my_threads = []

    def update_threads(self):
        my_threads_before = len(self.my_threads)
        for thread in self.my_threads:
            if not thread.is_alive():
                self.queues.remove(thread.queue)
        self.my_threads = [t for t in self.my_threads if t.is_alive()]
        if self.ws and my_threads_before > 0 and len(self.my_threads) == 0:
            if self.name == "BackChannel":
                print("All clients died. Stopping Stream for ", self.name)
            else:
                print("All clients died. Stopping Stream for ", self.name)

                msg = STOP_P2P_LIVESTREAM_MESSAGE.copy()
                msg["serialNumber"] = self.serialno
                asyncio.run(self.ws.send_message(json.dumps(msg)))

    def run(self):
        while self.run_event.is_set():
            self.update_threads()
            sys.stdout.flush()
            try:
                client_sock, client_addr = self.socket.accept()
                print ("New connection added: ", client_addr, " for ", self.name)
                sys.stdout.flush()

                if self.name == "BackChannel":
                    client_sock.setblocking(True)
                    print("Starting BackChannel")
                    thread = ClientRecvThread(client_sock, run_event, self.name, self.ws, self.serialno)
                    thread.start()
                else:
                    client_sock.setblocking(False)
                    thread = ClientSendThread(client_sock, run_event, self.name, self.ws, self.serialno)
                    self.queues.append(thread.queue)
                    if self.ws:
                        msg = START_P2P_LIVESTREAM_MESSAGE.copy()
                        msg["serialNumber"] = self.serialno
                        asyncio.run(self.ws.send_message(json.dumps(msg)))
                    self.my_threads.append(thread)
                    thread.start()
            except socket.timeout:
                pass

class ClientSendThread(threading.Thread):
    def __init__(self,client_sock,run_event,name,ws,serialno):
        threading.Thread.__init__(self)
        self.client_sock = client_sock
        self.queue = Queue(100)
        self.run_event = run_event
        self.name = name
        self.ws = ws
        self.serialno = serialno

    def run(self):
        try:
            while self.run_event.is_set():
                if self.queue.empty():
                    # Send something to know if socket is dead
                    self.client_sock.sendall(bytearray(0))
                    time.sleep(0.1)
                self.client_sock.sendall(
                    bytearray(self.queue.get(True)["data"])
                )
        except socket.error as e:
            print("Connection lost", self.name, e)
            pass
        except socket.timeout:
            print("Timeout on socket for ", self.name)
            pass
        self.client_sock.close()

class ClientRecvThread(threading.Thread):
    def __init__(self,client_sock,run_event,name,ws,serialno):
        threading.Thread.__init__(self)
        self.client_sock = client_sock
        self.run_event = run_event
        self.name = name
        self.ws = ws
        self.serialno = serialno

    def run(self):
        try:
            while self.run_event.is_set():
                try:
                    data = self.client_sock.recv(RECV_CHUNK_SIZE)
                    print("got ", len(data))
                except BlockingIOError:
                    # Resource temporarily unavailable (errno EWOULDBLOCK)
                    pass
        except socket.error as e:
            print("Connection lost", self.name, e)
            pass
        except socket.timeout:
            print("Timeout on socket for ", self.name)
            pass
        self.client_sock.close()
        msg = []
        msg["serialNumber"] = self.serialno
        asyncio.run(self.ws.send_message(json.dumps(msg)))

class Connector:
    def __init__(
        self,
        run_event,
    ):
        video_sock.bind(("127.0.0.1", 63336))
        video_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        video_sock.settimeout(1) # timeout for listening
        video_sock.listen()
        audio_sock.bind(("127.0.0.1", 63337))
        audio_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        audio_sock.settimeout(1) # timeout for listening
        audio_sock.listen()
        backchannel_sock.bind(("127.0.0.1", 63338))
        backchannel_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        backchannel_sock.settimeout(1) # timeout for listening
        backchannel_sock.listen()
        self.ws = None
        self.run_event = run_event
        self.serialno = ""

    def setWs(self, ws : EufySecurityWebSocket):
        self.ws = ws

    async def on_open(self):
        print(f" on_open - executed")

    async def on_close(self):
        print(f" on_close - executed")

    async def on_error(self, message):
        print(f" on_error - executed - {message}")

    async def on_message(self, message):
        #print(f"- on_message {message}")
        payload = message.json()
        message_type: str = payload["type"]
        if message_type == "result":
            message_id = payload["messageId"]
            print(f"on_message - {payload}")
            if message_id == START_LISTENING_MESSAGE["messageId"]:
                message_result = payload[message_type]
                states = message_result["state"]
                for state in states["devices"]:
                    self.serialno = state["serialNumber"]
                self.video_thread = ClientAcceptThread(video_sock, run_event, "Video", self.ws, self.serialno)
                self.audio_thread = ClientAcceptThread(audio_sock, run_event, "Audio", self.ws, self.serialno)
                self.backchannel_thread = ClientAcceptThread(backchannel_sock, run_event, "BackChannel", self.ws, self.serialno)
                self.audio_thread.start()
                self.video_thread.start()
                self.backchannel_thread.start()

        if message_type == "event":
            message = payload[message_type]
            event_type = message["event"]
            if message["event"] == "livestream audio data":
                event_value = message[EVENT_CONFIGURATION[event_type]["value"]]
                event_data_type = EVENT_CONFIGURATION[event_type]["type"]
                if event_data_type == "event":
                    for queue in self.audio_thread.queues:
                        if queue.full():
                            # print("Video queue full.")
                            queue.get(False)
                        queue.put(event_value)
            if message["event"] == "livestream video data":
                event_value = message[EVENT_CONFIGURATION[event_type]["value"]]
                event_data_type = EVENT_CONFIGURATION[event_type]["type"]
                if event_data_type == "event":
                    for queue in self.video_thread.queues:
                        if queue.full():
                            # print("Video queue full.")
                            queue.get(False)
                        queue.put(event_value)
            if message["event"] == "livestream error":
                print("Livestream Error!")
                if self.ws and len(self.video_thread.queues) > 0:
                    msg = START_P2P_LIVESTREAM_MESSAGE.copy()
                    msg["serialNumber"] = self.serialno
                    asyncio.run(self.ws.send_message(json.dumps(msg)))

def buildFfmpegCommand(cam_name):
    ffmpegcommand = [
        ffmpegbin,
        "-y",
#        "-analyzeduration", "1200000",
#        "-rtsp_transport", "tcp",
#        "-f", "h264", 
        "-i", "http://192.168.21.40:1997/api/frame.mp4?src=" + cam_name,
#        "-hls_init_time", "0",
#        "-hls_time", "2",
#        "-hls_segment_type", "mpegts",
#        "-fflags", "genpts+nobuffer+flush_packets",
#        "-frames:v", "1",
        "/config/www/eufyp2p/" + cam_name + ".jpg"
        ]
    return ffmpegcommand

def buildGstCmd(cam_name):
    gstcommand = [
        "gst-launch-1.0",
        "-v",
#        "rtspsrc",
#        "protocols=tcp",
#        "location=rtsp://127.0.0.1:" + str(RTSP_PORT_NUMBER) + "/" + cam_name + "?mp4",
#        "latency=0",
        "souphttpsrc",
        "location=http://192.168.21.40:1997/api/frame.mp4?src=" + cam_name,
        "!",
        "decodebin",
        "!",
        "jpegenc",
        "snapshot=true",
        "!",
        "filesink",
        "location=/config/www/eufyp2p/" + cam_name + ".jpg"
        ]
    return gstcommand

# build as non-async function and call in main
# maybe... build as async with subprocess.returncode used with await, build loop outside of function call
#async def snapshotCmd(cam_name, snapshot_interval):
#    ffmpegcmd = buildFfmpegCommand(cam_name)
#    print(ffmpegcmd)
#    await asyncio.sleep(60)
#    while True:
#        print("Snapshot starting")
#        subprocess.Popen(ffmpegcmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
#        print("Snapshot created")
#        await asyncio.sleep(snapshot_interval)
        #else:
            #print("Error during snapshot, retrying in 10 seconds")
            #await asyncio.sleep(10)


async def main(run_event):

    SSINTERVAL = 1800 # set as default, get envs not working returned none
    audioVideoStream = 'camera1' # set default, placeholder till sed for go2rtc made
    videoStream = 'camera1_video'
    audioStream = 'camera1_audio'
    c = Connector(run_event)

    ws: EufySecurityWebSocket = EufySecurityWebSocket(
        "127.0.0.1",
        3000,
        aiohttp.ClientSession(),
        c.on_open,
        c.on_message,
        c.on_close,
        c.on_error,
    )
    c.setWs(ws)
    await ws.connect()

    await ws.send_message(json.dumps(START_LISTENING_MESSAGE))
    await ws.send_message(json.dumps(SET_API_SCHEMA))
    await ws.send_message(json.dumps(DRIVER_CONNECT_MESSAGE))

    print("Main stream name is: ")
    print(audioVideoStream)
    print("Video only stream name is: ")
    print(videoStream)
    print("Snapshot interval is: ")
    print(SSINTERVAL)
    snapcmd = buildFfmpegCommand(videoStream)
    print("Snapshot command is:")
    print(" ".join(snapcmd))
    await asyncio.sleep(20)
    while True: #need use ifs with returncode
        print("Snapshot snapping")
        subprocess.Popen(snapcmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        await asyncio.sleep(SSINTERVAL)
    await asyncio.sleep(1000000000000000000000005)

try:
    asyncio.run(main(run_event))

except (KeyboardInterrupt, SystemExit):
    run_event.clear()
